File: csharp/20_modules.cs
// ============================================================
// Topic:  Modules & Namespaces
// Summary: Namespaces, using System.Math, using alias, block namespace.
// Run:     dotnet script csharp/20_modules.cs
//
// KEY DIFFERENCES:
//   C#:     namespace, using, file-scoped namespace, global using
//   Python: import module, from x import y, as alias
//   TS/JS:  import/export, ES modules vs CommonJS
//   Rust:   mod, pub, use, crate
//   Java:   package, import
//
// NOTE: dotnet script disallows inline `namespace` (CS7021).
//       Nested static types below mirror block namespaces; in a .csproj:
//         namespace MyApp.Constants { public static class Pi { ... } }
//         namespace MyApp.Greetings;  // file-scoped (C# 10+)
//         global using System;        // GlobalUsings.cs in SDK projects
// ============================================================

using System;
using Greeter = MyApp.Greetings.Greeter;           // type alias| Python: from myapp.greetings import Greeter
using Pow = System.Math;                           // alias     | Python: import math as pow

// ----- block namespace equivalents (nested static types for dotnet script) -----
static class MyApp
{
    public static class Constants
    {
        public static class Pi
        {
            public static double Value => 3.14;    // constant  | demo value (not Math.PI)
        }
    }

    public static class Greetings
    {
        public static class Greeter
        {
            public static string Greet() => "Hi";  // greet     | callable from main via using
        }
    }
}

// ----- top-level main — uses Math, namespaces, and alias -----
Console.WriteLine($"pi: {MyApp.Constants.Pi.Value:F2}");  // pi: 3.14 | MyApp.Constants
Console.WriteLine($"greet: {Greeter.Greet()}");           // greet: Hi | using Greeter = MyApp.Greetings.Greeter
Console.WriteLine($"sqrt: {Math.Sqrt(25):F1}");           // sqrt: 5.0 | System.Math
Console.WriteLine($"alias: {Pow.Pow(5, 2)}");             // alias: 25 | Pow aliases System.Math
Console.WriteLine("main module");                         // main module | entry-point output

// Cross-language note:
//   file-scoped namespace (C# 10+): `namespace MyApp;` — entire file in one namespace.
//   global using in GlobalUsings.cs — project-wide imports (SDK-style projects).
//   using static System.Math — import static members (Sqrt, PI) without type prefix.
//   Namespaces are logical grouping only — unlike Rust/Java packages, not physical modules by default.

File: rust/20_modules.rs
// ============================================================
// Topic:  Modules
// Summary: mod, pub, use, inline submodule, re-export alias, std imports.
// Run:     rustc --edition 2021 rust/20_modules.rs -o rust/20_modules.exe && rust/20_modules.exe
//
// KEY DIFFERENCES:
//   Python:  import / from … import; __init__.py packages
//   TS/JS:   import / export; ES modules vs CommonJS
//   Rust:    mod tree, pub for visibility, use for paths; one crate root (main.rs)
//   Java:    package + import
//   C++:     #include / modules (C++20)
// ============================================================

use std::f64::consts::PI;

// ----- 1. Inline submodule — nested mod with pub fn -----
mod submodule {
    pub fn greet() -> &'static str {
        "Hi" // pub = visible outside this mod; default is private
    }
}

// ----- 2. Inline math_util module -----
mod math_util {
    pub fn sqrt_val(x: f64) -> f64 {
        x.sqrt()
    }

    pub fn square(x: i32) -> i32 {
        x * x
    }
}

// ----- 3. use — bring names into scope -----
use submodule::greet;
use math_util::sqrt_val as sqrt; // alias: sqrt maps to sqrt_val
use math_util::square as alias;  // alias: square imported under another name

fn main() {
    // ----- 1. std constant via use -----
    println!("pi: {:.2}", PI); // Output: pi: 3.14   | Python: from math import pi

    // ----- 2. pub fn from submodule -----
    println!("greet: {}", greet()); // Output: greet: Hi

    // ----- 3. math_util via aliased import -----
    println!("sqrt: {:.1}", sqrt(25.0)); // Output: sqrt: 5.0

    // ----- 4. alias import — square renamed to alias -----
    println!("alias: {}", alias(5)); // Output: alias: 25

    // ----- 5. main module entry point -----
    println!("main module"); // Output: main module
    // Cross-language: Rust modules are a single tree per crate; no separate files required here.
    // In larger projects: mod math_util; → math_util.rs or math_util/mod.rs
}